### ISSUE TITLE
[JENKINS-40787] Lockable Resource Lock with inversePrecedence=true not granting lock to newest build

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/lockable-resources-plugin#861]
<p>Lockable resources with inversePrecedence=true is not functioning correctly when used in the steps block of a Declarative Pipeline. The locked resource is locked correctly but when released it does not grant the lock to the newest build but instead the first one that requested it, as if inversePrecedence=false. </p>

<p>I discovered that if you use the lock in the when block of the Declarative Pipeline the inversePrecedence behaves as expected. </p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=chris_carson">chris_carson</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40787" target="_blank">Lockable Resource Lock with inversePrecedence=true not granting lock to newest build</a></i></summary>
<i><ul>
<li><b>status</b>: Open
<li><b>priority</b>: Major
<li><b>component(s)</b>: lockable-resources-plugin
<li><b>label(s)</b>: pipeline
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 2
<li><b>watchers</b>: 3
<li><b>imported</b>: 20251217-081411
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Lockable resources with inversePrecedence=true is not functioning correctly when used in the steps block of a Declarative Pipeline. The locked resource is locked correctly but when released it does not grant the lock to the newest build but instead the first one that requested it, as if inversePrecedence=false. </p>

<p>I discovered that if you use the lock in the when block of the Declarative Pipeline the inversePrecedence behaves as expected. </p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote link associated with this issue:_


  - [CloudBees Internal CD-50](https://cloudbees.atlassian.net/browse/CD-50)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stefanthurnherr.png" title="stefanthurnherr's avatar" /> stefanthurnherr:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40787?focusedId=283403&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-283403">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Same problem discovered in our multi-branch Jenkinsfile pipeline (not in Declarative Pipeline context but same problem I guess...). The <tt>inversePrecedence:true</tt> doesn't seem to work reliably. Specifically, I've tried to protect our long-running integration tests with a LIFO queue using the following pipeline script:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">[...] <span class="code-comment">// other stages to build, <span class="code-keyword">static</span> code analysis, unit tests etc
</span>stage(<span class="code-quote">'Integration testing'</span>) {
        milestone(label: <span class="code-quote">'IntegrationTests-AttemptingToStart'</span>)
        <span class="code-object">String</span> lockResource = <span class="code-quote">"integrationTests-${env.BRANCH_NAME}"</span>
        lock(resource: lockResource, inversePrecedence: <span class="code-keyword">true</span>) {
            <span class="code-keyword">try</span> {
                integrationTestsStage() <span class="code-comment">//runs different types of integrationTests in parallel, using more lock()s <span class="code-keyword">for</span> protecting DB resource
</span>            } <span class="code-keyword">finally</span> {
                echo <span class="code-quote">"Done with all integration tests."</span>
                milestone(label:<span class="code-quote">'IntegrationTests-Done'</span>) <span class="code-comment">// abort all older builds (<span class="code-keyword">for</span> same git branch) that haven't passed <span class="code-keyword">this</span> milestone yet
</span>            }
        }
}
</pre>
</div></div>
<p>It did not work, the builds entered the integrationTests stage in strange orders and sometimes even two builds at the same time.</p>

<p><b>Update a few hours later:</b>
After some more investigation I think the problem is that one cannot have nested lock()s in a pipeline. In our case above, the parallel steps run within <tt>integrationTestsStage</tt> each hold an additional resource (DB) lock. But whenever one of the inner locks is released, all other locks held by the same build also get released.</p>

<p>I believe this is not only when using inversePrecedence, so may be a different bug?</p>

<p>For the reference: Our environment is Jenkins 2.37 with latest versions of all pipeline-related plugins </p>
</pre>
</details>

</ul>
</details>

<p>Same problem discovered in our multi-branch Jenkinsfile pipeline (not in Declarative Pipeline context but same problem I guess...). The <tt>inversePrecedence:true</tt> doesn't seem to work reliably. Specifically, I've tried to protect our long-running integration tests with a LIFO queue using the following pipeline script:</p>

<pre>
[...] <span class="code-comment">// other stages to build, <span class="code-keyword">static</span> code analysis, unit tests etc
</span>stage(<span class="code-quote">'Integration testing'</span>) {
milestone(label: <span class="code-quote">'IntegrationTests-AttemptingToStart'</span>)
<span class="code-object">String</span> lockResource = <span class="code-quote">"integrationTests-${env.BRANCH_NAME}"</span>
lock(resource: lockResource, inversePrecedence: <span class="code-keyword">true</span>) {
    <span class="code-keyword">try</span> {
integrationTestsStage() <span class="code-comment">//runs different types of integrationTests in parallel, using more lock()s <span class="code-keyword">for</span> protecting DB resource
</span>    } <span class="code-keyword">finally</span> {
echo <span class="code-quote">"Done with all integration tests."</span>
milestone(label:<span class="code-quote">'IntegrationTests-Done'</span>) <span class="code-comment">// abort all older builds (<span class="code-keyword">for</span> same git branch) that haven't passed <span class="code-keyword">this</span> milestone yet
</span>    }
}
}
</pre>
<p>It did not work, the builds entered the integrationTests stage in strange orders and sometimes even two builds at the same time.</p>

<p><b>Update a few hours later:</b><br/>
After some more investigation I think the problem is that one cannot have nested lock()s in a pipeline. In our case above, the parallel steps run within <tt>integrationTestsStage</tt> each hold an additional resource (DB) lock. But whenever one of the inner locks is released, all other locks held by the same build also get released.</p>

<p>I believe this is not only when using inversePrecedence, so may be a different bug?</p>

<p>For the reference: Our environment is Jenkins 2.37 with latest versions of all pipeline-related plugins </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>chris_carson:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40787?focusedId=283770&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-283770">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stefanthurnherr" class="user-hover" rel="stefanthurnherr">stefanthurnherr</a> In my case I did not have nested locks. So I would create a new separate issue for that.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stefanthurnherr" class="user-hover" rel="stefanthurnherr">stefanthurnherr</a> In my case I did not have nested locks. So I would create a new separate issue for that.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stefanthurnherr.png" title="stefanthurnherr's avatar" /> stefanthurnherr:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40787?focusedId=283919&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-283919">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Done: <a href="https://issues.jenkins.io/browse/JENKINS-41183" title="Pipeline/Jenkinsfile: Nested lock() statements not working as expected" class="issue-link" data-issue-key="JENKINS-41183"><del>JENKINS-41183</del></a></p>
</pre>
</details>

</ul>
</details>

<p>Done: <a href="https://issue-redirect.jenkins.io/issue/41183" title="Pipeline/Jenkinsfile: Nested lock() statements not working as expected" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/41183">JENKINS-41183</a>"><del>JENKINS-41183</del></a></p>
